runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -137,12 +137,32 @@ private void processElementWithSdkHarness(WindowedValue<InputT> element) throws
     checkState(
         stateRequestHandler != null, "%s not yet prepared", StateRequestHandler.class.getName());
 
+    final ArrayList<KV<String, OutputT>> results = new ArrayList();
+
+    OutputReceiverFactory receiverFactory =
+        new OutputReceiverFactory() {
+          @Override
+          public FnDataReceiver<OutputT> create(String pCollectionId) {
+            return (receivedElement) -> {
+              synchronized (results) {
+                // buffer to not block the grpc thread
+                results.add(KV.of(pCollectionId, receivedElement));
+              }
+            };
+          }
+        };
+
     try (RemoteBundle bundle =
-        stageBundleFactory.getBundle(
-            new ReceiverFactory(outputManager, outputMap), stateRequestHandler, progressHandler)) {
+        stageBundleFactory.getBundle(receiverFactory, stateRequestHandler, progressHandler)) {
       logger.debug(String.format("Sending value: %s", element));
       // TODO(BEAM-4681): Add support to Flink to support portable timers.
       Iterables.getOnlyElement(bundle.getInputReceivers().values()).accept(element);
+      // TODO: it would be nice to emit results as they arrive, can thread wait non-blocking?
+    }
+
+    // RemoteBundle close blocks until all results are received
+    for (KV<String, OutputT> result : results) {
```

Comment:
We can use threadpool to process these elements instead of buffering all the elements.

Reply:
Flink operators (and probably their entire chain) cannot be assumed to be thread-safe. The  best solution would be to use the thread that is blocking for the end of the bundle. The next best thing would be a single output background thread and synchronize after the bundle is closed. At this time that isn't the top priority though..